### PR TITLE
Add loop related optimizations to LLVM JIT

### PR DESCRIPTION
**Pull request by EgorBo**
Example:
```csharp
[MethodImpl(MethodImplOptions.NoInlining)]
static long Test(long x, bool condition)
{
    for (int i=10; i<100000; i+=2)
    {
        if (condition)
            x++;
    }
    return x;
}
```
### Before my change:
LLVM IR:
```llvm
define i64 @"Program:Test (long,bool)"(i64 %arg_x, i32 %arg_condition) {
BB0:
  br label %BB3
BB3:                                              ; preds = %BB0
  br label %BB2
BB2:                                              ; preds = %BB3
  br label %BB4
BB4:                                              ; preds = %BB6, %BB2
  %0 = phi i32 [ 10, %BB2 ], [ %t37, %BB6 ]
  %1 = phi i64 [ %arg_x, %BB2 ], [ %4, %BB6 ]
  %2 = icmp slt i32 %0, 100000
  br i1 %2, label %BB5, label %BB8
BB5:                                              ; preds = %BB4
  %3 = icmp eq i32 %arg_condition, 0
  br i1 %3, label %BB6, label %BB7
BB8:                                              ; preds = %BB4
  br label %BB1
BB6:                                              ; preds = %BB7, %BB5
  %4 = phi i64 [ %1, %BB5 ], [ %t35, %BB7 ]
  %t37 = add i32 %0, 2
  br label %BB4
BB7:                                              ; preds = %BB5
  %t35 = add i64 %1, 1
  br label %BB6
BB1:                                              ; preds = %BB8
  ret i64 %1
}
```
ASM:
```asm
  mov eax, 10
  cmp eax, 99999
  jg .LBB0_3
.LBB0_2: 
  cmp esi, 1
  sbb rdi, -1
  add eax, 2
  cmp eax, 99999
  jle .LBB0_2
.LBB0_3: 
  mov rax, rdi
  ret
```
### After my change:
LLVM IR:
```llvm
define i64 @"Program:Test (long,bool)"(i64 %arg_x, i32 %arg_condition) {
  %0 = icmp eq i32 %arg_condition, 0
  %1 = select i1 %0, i64 0, i64 49995
  %2 = add i64 %1, %arg_x
  ret i64 %2
}
```
ASM:
```asm
  xor ecx, ecx
  test esi, esi
  mov eax, 49995
  cmove rax, rcx
  add rax, rdi
  ret
```

I spent some time analyzing the "Before/After" diffs generated by `opt -O1` for our IR: https://godbolt.org/z/vRCpS3 in order to compose this set.
Some `dotnet/performance` benchmarks use such loops (with a constant) 

### Diffs:
  
`-simplifycfg`   [diff](https://www.diffchecker.com/Uy5Gs7ah)
`-sroa` (no changes)
`-early-cse` (no changes)
`-instcombine`   [diff](https://www.diffchecker.com/26Wy6VxP)
`-correlated-propagation`   [diff](https://www.diffchecker.com/41SqRtfP)
`-instcombine`   [diff](https://www.diffchecker.com/Z9R8f8n1)
`-loop-simplify`  (no changes)
`-lcssa`   [diff](https://www.diffchecker.com/NW9LIjnc)
`-loop-rotate`   [diff](https://www.diffchecker.com/GtbOu09E)
`-licm`   [diff](https://www.diffchecker.com/rQmNJx0w)
`-instcombine`   [diff](https://www.diffchecker.com/vdLZgmKs)
`-lcssa`   [diff](https://www.diffchecker.com/yuqVwXdc)
`-indvars`   [diff](https://www.diffchecker.com/ihEzoBTF)
`-loop-deletion`   [diff](https://www.diffchecker.com/EJT6aLfI)
 `-instcombine`   [diff](https://www.diffchecker.com/qMOqH7dt)
(~~TODO: add~~ `-simplifycfg` at the end to remove that redundant BB0 after previous pass) - [diff](https://www.diffchecker.com/flsgO9S8)

### CODE REVIEW

File: mono/mini/llvm-jit.cpp [hunk 1]
```diff
@@ -288,7 +288,7 @@ class MonoLLVMJIT {
 		if (opts == NULL) {
 			// FIXME: find optimal mono specific order of passes
 			// see https://llvm.org/docs/Frontend/PerformanceTips.html#pass-ordering
-			opts = " -simplifycfg -sroa -instcombine -gvn";
+			opts = " -simplifycfg -sroa -early-cse -instcombine -correlated-propagation -instcombine -loop-simplify -lcssa -loop-rotate -licm -instcombine -lcssa -indvars -loop-deletion -instcombine -simplifycfg";
```
Comment (filipnavara) [inline mono/mini/llvm-jit.cpp]: Nit: simplifycfg is there twice

Comment (EgorBo) [reply mono/mini/llvm-jit.cpp]: @filipnavara the [last ](https://www.diffchecker.com/flsgO9S8)`-simplifycfg` removes a redundant block generated by previous pass (instcombine). The [first ](https://www.diffchecker.com/Uy5Gs7ah)`-simplifycfg` removes redundant blocks generated by mono 🙂

**Comment by EgorBo** [comment]
I am working on some jit diff tools, e.g. the following benchmark: 
https://github.com/dotnet/performance/blob/7d96fd3dea66b83a0b02222e44e02e762a990c32/src/benchmarks/micro/corefx/System.Memory/ReadOnlySpan.cs#L39
Generates: https://www.diffchecker.com/SjgMjh3n diff

**Comment by EgorBo** [comment]
It seems we should also add the `Lower 'expect' Intrinsics` pass (to convert our llvm.expect calls into weights)

File: mono/mini/llvm-jit.cpp (hunk 1 above)
Comment (Therzok) [reply mono/mini/llvm-jit.cpp]: @EgorBo Maybe it's worth adding a comment in the code so nobody removes them in the future?

**Comment by lewurm** [comment]
@monojenkins build failed

**Comment by lewurm** [comment]
@monojenkins squash

**Comment by EgorBo** [comment]
I am not 100% in this set yet, need to run the benchmarks and diffs one more time to make sure it doesn't regress anything. (however some passes are indeed should be added)

**Comment by lewing** [comment]
Looks like this needs a conflict resolved

**Comment by marek-safar** [comment]
@EgorBo what is the plan with this PR?

**Comment by EgorBo** [comment]
@marek-safar yeah will update it soon, this one have two mistakes:
`-early-cse` for some reason slows down benchmarks - need to figure out why
`-ipsccp` should be replaced with `-sccp` (ipsccp is interproc)

**Comment by lewing** [comment]
Checking in again.

**Comment by EgorBo** [comment]
Just updated the pass list.
My algorithm:
Run a benchmark in LLVM AOT mode with `-print-before-all`, `-print-after-all` and look for passes with useful changes -- make a list of those (some common sequence):

Benchmarks (the new pass order is a baseline):
image

(benchmarks: https://github.com/nxrighthere/BurstBenchmarks)

`dotnet/benchmarks`: mostly positive effect, a few 5-10% regressions:
image

**Comment by EgorBo** [comment]
@monojenkins build failed

**Pull request merged**